File: 416-partition-equal-subset-sum/partition-equal-subset-sum.cpp
class Solution {
public:
      bool f(int idx,vector<int>&arr, int target, vector<vector<int>>&dp){
        if(target==0)return dp[idx][target]=true;
        if(idx==0)return dp[idx][target]=arr[idx]==target;
       
        if(dp[idx][target]!=-1)return dp[idx][target];
        bool not_pick=f(idx-1,arr,target,dp);
        bool pick=false;
        if(arr[idx]<=target){
           pick=f(idx-1,arr,target-arr[idx],dp);
        }
            return dp[idx][target]=pick|not_pick;
    }
    bool canPartition(vector<int>& nums) {
        int sum=0;
        for(auto it:nums)sum+=it;
        if(sum%2==1)return 0;
        sum=sum/2;
         vector<vector<int>>dp(nums.size(),vector<int>(sum+1,-1));
        return f(nums.size()-1,nums,sum,dp);
    }
};

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.java
class Solution {
    boolean f(int idx, int[] nums, int target, ArrayList<ArrayList<Boolean>>dp){
        if(idx==nums.length)return false;
        if(target==0)return true; 

        if(dp.get(idx).get(target)!=null)return dp.get(idx).get(target);
       
        boolean take=false;
        if(target>=nums[idx])take=f(idx+1,nums, target-nums[idx],dp);

        boolean not_take=f(idx+1,nums,target,dp);

        dp.get(idx).set(target,take | not_take);

        return take | not_take;
    }
    public boolean canPartition(int[] nums) {
        int sum=Arrays.stream(nums).sum();
        if(sum%2==1)return false;
        int target=sum/2;
        ArrayList<ArrayList<Boolean>>dp=new ArrayList<>();

        for(int i=0;i<nums.length;i++){
             dp.add(new ArrayList<>(Collections.nCopies(target + 1, null)));
        }

        return f(0,nums,target,dp);
    }
}